File: Shared/Models/TicketPlatform.swift
import Foundation

enum TicketPlatform: String, CaseIterable, Sendable {
    case ePlus
    case ticketPia
    case lawsonTicket
    case livePocket
    case tiget
    case zaiko

    var ticketAccessURL: URL {
        switch self {
        case .ePlus:
            URL(string: "https://eplus.jp/jyoukyou/")!
        case .ticketPia:
            URL(string: "https://ticket-account.pia.jp/pia/digipoke/list.do")!
        case .lawsonTicket:
            URL(string: "https://l-tike.com/mypage/")!
        case .livePocket:
            URL(string: "https://t.livepocket.jp/my_ticket")!
        case .tiget:
            URL(string: "https://tiget.net/users/sign_in")!
        case .zaiko:
            URL(string: "https://zaiko.io/account/eticket")!
        }
    }

    static func detect(from purchaseURL: URL) -> TicketPlatform? {
        guard let host = purchaseURL.host(percentEncoded: false)?.lowercased() else {
            return nil
        }

        return allCases.first { platform in
            platform.hosts.contains { host == $0 || host.hasSuffix(".\($0)") }
        }
    }

    private var hosts: [String] {
        switch self {
        case .ePlus:
            ["eplus.jp"]
        case .ticketPia:
            ["pia.jp"]
        case .lawsonTicket:
            ["l-tike.com"]
        case .livePocket:
            ["livepocket.jp"]
        case .tiget:
            ["tiget.net"]
        case .zaiko:
            ["zaiko.io"]
        }
    }
}

File: OshiLifeAndroid/lib/data/models/ticket_platform.dart
/// Port of `Shared/Models/TicketPlatform.swift`.
///
/// Detection is dot-anchored host matching: `host == h` or
/// `host.endsWith('.' + h)` — so `eplus.jp.example.com` must NOT match.
enum TicketPlatform {
  ePlus,
  ticketPia,
  lawsonTicket,
  livePocket,
  tiget,
  zaiko;

  Uri get ticketAccessUrl => switch (this) {
    ePlus => Uri.parse('https://eplus.jp/jyoukyou/'),
    ticketPia => Uri.parse(
      'https://ticket-account.pia.jp/pia/digipoke/list.do',
    ),
    lawsonTicket => Uri.parse('https://l-tike.com/mypage/'),
    livePocket => Uri.parse('https://t.livepocket.jp/my_ticket'),
    tiget => Uri.parse('https://tiget.net/users/sign_in'),
    zaiko => Uri.parse('https://zaiko.io/account/eticket'),
  };

  List<String> get _hosts => switch (this) {
    ePlus => const ['eplus.jp'],
    ticketPia => const ['pia.jp'],
    lawsonTicket => const ['l-tike.com'],
    livePocket => const ['livepocket.jp'],
    tiget => const ['tiget.net'],
    zaiko => const ['zaiko.io'],
  };

  static TicketPlatform? detect(Uri purchaseUrl) {
    final host = purchaseUrl.host.toLowerCase();
    if (host.isEmpty) return null;
    for (final platform in values) {
      for (final candidate in platform._hosts) {
        if (host == candidate || host.endsWith('.$candidate')) {
          return platform;
        }
      }
    }
    return null;
  }
}
